rules/prefer-error-cause.js

```diff
@@ -0,0 +1,341 @@
+		result.push({
+			node: errorConstructorLastArgument,
+			statementToFix: throwStatement,
+		});
+	} else {
+		// Assume Error's constructor exists in other node of the block
+		// It could be VariableDeclarator or AssignmentExpression.
+
+		const newExpressionSelector = reference =>
+			(reference.identifier?.parent.type === 'AssignmentExpression'
+			|| reference.identifier?.parent.type === 'VariableDeclarator')
+			&& reference.writeExpr?.type === 'NewExpression'
+			&& reference.resolved?.name === throwStatement.argument.name;
+
+		const thrownErrorDeclarators = getReferences(context.getScope())
+			.filter(reference => newExpressionSelector(reference))
+			.map(reference => reference.identifier.parent);
+
+		// Report newExpression not found in given block.
+		// `try {} catch { const err = new Error; try {} catch { throw err; } }`
+		if (_.isEmpty(thrownErrorDeclarators)) {
+			reportCannotBeFixed(node, context);
+			return;
+		}
+
+		// `try {} catch (err) { let e1 = new Error('oops'); e1 = new Error('oops'); throw error;}`
+
+		for (const thrownErrorDeclarator of thrownErrorDeclarators) {
+			statementToFix = thrownErrorDeclarator;
+
+			let targetArguments;
+			if (thrownErrorDeclarator.type === 'VariableDeclarator') {
+				targetArguments = thrownErrorDeclarator.init.arguments;
+			} else if (thrownErrorDeclarator.type === 'AssignmentExpression') {
+				targetArguments = thrownErrorDeclarator.right.arguments;
+			}
+
+			errorConstructorLastArgument = peekTop(targetArguments);
+
+			if (!errorConstructorLastArgument) {
+				reportCannotBeFixed(node, context);
+				return;
+			}
+
+			result.push({
+				node: errorConstructorLastArgument,
+				statementToFix,
+			});
+		}
+	}
+
+	return result;
+};
+
+const handleCatchBlock = ({node, context, statements, parameter}) => {
+	const throwStatements = esquery.match(node, esquery.parse('ThrowStatement'), statements);
+
+	for (const throwStatement of throwStatements) {
+		// Filter blocks not having throw statement.
+		// `try {} catch (error) {}`
+		// `promise.then(undefined, error => {})`
+		// `promise.catch(error => {})`
+		if (!throwStatement) {
+			continue;
+		}
+
+		const errorArgumentIdentifier = parameter?.name;
+		const throwStatementArgument = throwStatement.argument;
+
+		// Filter rethrowing the error itself.
+		// `try {} catch (err) { throw err; }`
+		// `promise.then(undefined, (err) => { throw err; })`
+		// `promise.catch(err => { throw err; })`
+		if (errorArgumentIdentifier && errorArgumentIdentifier === throwStatementArgument.name) {
+			continue;
+		}
+
+		// Report none identifier parameter
+		// `try {} catch ({error}) {}`
+		// `promise.then(undefined, ({error}) => {})`
+		// `promise.catch({error} => {})`
+		if (parameter && parameter.type !== 'Identifier') {
+			context.report({
+				node,
+				messageId: ERROR,
+			});
+			continue;
+		}
+
+		const nodesToFixCandidates = getAllNodesToFix({node, context, throwStatement});
+		if (_.isEmpty(nodesToFixCandidates)) {
```

Comment:
Empty array can also filter and `for...of`, don't really see the point checking empty array here.